### PR TITLE
`ActorCriticRLModel`: Don't unnecessarily reinitialize Runner

### LINKED ISSUE #619: Support learn() with total timesteps less than episode length
In an imitation learning project, I'm alternating short calls (low `total_timesteps`) to `PPO2.learn()` with gradient descent updates to the discriminator among other things.

The PPO2 updates themselves seem to be working fine. However, logging has a few problems because some logging state is not kept between different calls to `learn()`. Despite the fact that I use `.learn(reset_num_timesteps=False)`, which seems to exist to allow logging between multiple calls to `PPO2.learn()`, we run these problems:

(1) We initialize a new Runner every time we call `learn`, even if `reset_num_timesteps=False`. This forces the environment to reset(), thus biasing the logger towards reporting the reward and episode length means of shorter episodes (longer episodes don't get to finish by the end of the training loop). As an example, when training CartPole, my expert imitation policies (mean return: 500) often show mean return around 300 in my PPO2 logs.

(2) The `ep_info_buf` is reset every time we call learn instead of being saved. This makes the training curve more jagged and again biased toward shorter episodes at the beginning of each new call to `learn()`.

I'm wondering if the maintainers would be interested in some sort of PR that fixes these logging discrepancies.

I propose that `reset_num_timesteps=False` (perhaps rename to `reset_log_state`) should make PPO2 keep state from (1) and (2).

### PR DESCRIPTION
**Pull request by shwang**
## Description
Main change: 
  * Fixes a bug in PPO2, ACER, A2C, and ACKTR where repeated calls to `learn(total_timesteps)` reset the environment on every call, potentially biasing samples toward early episode timesteps.
  * Fixed by adding lazy property `ActorCriticRLModel.runner`. Subclasses now use lazily-generated `self.runner` instead of reinitializing a new Runner every time `learn()` is called.

Other changes:
  * Used `self.n_{act,batch}` over `self.runner.n_{act,batch}` in `acer_simple.py` to avoid a pytype "attribute not found" error due to the type of `self.runner` being `AbstractRunner` which doesn't have `self.n_batch`.
  * Refactored redundant parts of `common/identity_env.py` and added a `num_resets` attribute that was helpful for making tests.
  * In an earlier iteration of this PR, `self.runner` was not lazily generated (instead it was generated by `ActorCriticRLModel.__init__()`) and this required that `super().__init__()` inside subclasses of `ActorCriticRLModel` be called after certain attributes are initialized inside `Subclass.__init__()`.
    * I moved the `super().__init__()` statement to the end of `Subclass.__init__()`.
    * I removed some redundant attribute initialization `self.X = None` inside `Subclass.__init__()` when `super().__init__()` did this already with another `self.X = None` statement.
    * I'm ok with removing this refactor change since it's not related to the original issue.

## Motivation and Context

addresses part of #619 (doesn't provide fix for OffPolicyRLModel, MPI-based models)
- [x] I have raised an issue to propose this change ([required](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the [changelog](https://github.com/hill-a/stable-baselines/blob/master/docs/misc/changelog.rst) accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [x] I have ensured `pytest` and `pytype` both pass.

### CODE REVIEW

File: stable_baselines/acktr/acktr.py [hunk 1]
```diff
@@ -115,9 +102,26 @@ def __init__(self, policy, env, gamma=0.99, nprocs=None, n_steps=20, ent_coef=0.
         self.trained = False
         self.continuous_actions = False
 
+        super(ACKTR, self).__init__(policy=policy, env=env, verbose=verbose, requires_vec_env=True,
+                                    _init_setup_model=_init_setup_model, policy_kwargs=policy_kwargs,
+                                    seed=seed, n_cpu_tf_sess=n_cpu_tf_sess)
+
+        if nprocs is not None:
```
Comment (AdamGleave) [inline stable_baselines/acktr/acktr.py]: ibid: why are we moving input validation later?

File: tests/test_multiple_learn.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    # TD3,
+]
+
+
+@pytest.mark.parametrize("model_class", MODEL_LIST)
+def test_model_multiple_learn_no_reset(model_class):
+    """Check that when we call learn multiple times, we don't unnecessarily
+    reset the environment.
+    """
+    if model_class is ACER:
```
Comment (AdamGleave) [inline tests/test_multiple_learn.py]:
```suggestion
    if model_class is ACER:  # only supports discrete actions
```

File: tests/test_multiple_learn.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        def make_env():
+            return IdentityEnv(ep_length=1e10, dim=2)
+    else:
+        def make_env():
+            return IdentityEnvBox(ep_length=1e10)
+    venv = DummyVecEnv([make_env])
```
Comment (AdamGleave) [inline tests/test_multiple_learn.py]: Might be simpler to create `env` and then embed it in `venv` given we need to retrieve it anyway? e.g:

```
env = make_env()
venv = DummyVecEnv([lambda: env])
```

File: tests/test_multiple_learn.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    env = venv.envs[0]
+    assert env.num_resets == 0
+
+    model = model_class(policy="MlpPolicy", env=venv)
+    assert env.num_resets == 0
+    for _ in range(4):
+        model.learn(total_timesteps=400)
```
Comment (AdamGleave) [inline tests/test_multiple_learn.py]: Maybe also directly check the `runner` property hasn't changed (same object)? It's conceivable (although unlikely given current implementation) that the runner changes, does not reset, but that some runner-specific stats still get lost for example.

File: stable_baselines/common/base_class.py [hunk 5]
```diff
@@ -736,6 +737,25 @@ def __init__(self, policy, env, _init_setup_model, verbose=0, policy_base=ActorC
         self.step = None
         self.proba_step = None
         self.params = None
+        self._runner = None
+
+    def _make_runner(self) -> AbstractEnvRunner:
```
[the remark this answers is not in the document]
Comment (shwang) [reply stable_baselines/common/base_class.py]: No, because TRPO/PPO1 don't use a runner but extend this class.

File: stable_baselines/common/base_class.py [hunk 6]
```diff
@@ -736,6 +737,25 @@ def __init__(self, policy, env, _init_setup_model, verbose=0, policy_base=ActorC
...
+
+    @property
+    def runner(self) -> AbstractEnvRunner:
+        if self._runner is None:
+            self._runner = self._make_runner()
+        return self._runner
+
+    def set_env(self, env):
+        if self.env is not env:
+            self._runner = None  # New environment invalidates `self._runner`.
```
[the remark this answers is not in the document]
Comment (shwang) [reply stable_baselines/common/base_class.py]: I'm 90% sure that this used to just be garbage collected after every `.learn()`, but I will check to make sure

**Comment by shwang** [comment]
I'm removing the `IdentityBox.dim` attribute -- nothing is using it in our codebase and I'm guessing no else uses it either, if there's anyone using IdentityBox at all.

Main reason I'm removing is because don't want to infer `dim` (always an integer, and now is an optional parameter to `IdentityBox.__init__`) from `space.shape` (sometimes an integer, sometimes tuple).

File: stable_baselines/acer/acer_simple.py [hunk 7]
```diff
@@ -135,36 +131,39 @@ def __init__(self, policy, env, gamma=0.99, n_steps=20, num_procs=None, q_coef=0
...
         self.n_act = None
         self.n_batch = None
         self.summary = None
         self.episode_reward = None
 
+        super(ACER, self).__init__(policy=policy, env=env, verbose=verbose, requires_vec_env=True,
+                                   _init_setup_model=_init_setup_model, policy_kwargs=policy_kwargs,
+                                   seed=seed, n_cpu_tf_sess=n_cpu_tf_sess)
+
+        if num_procs is not None:
```
[the remark this answers is not in the document]
Comment (shwang) [reply stable_baselines/acer/acer_simple.py]: It's important that this if block occurs after `super.__init__()` because `BaseRLModel.__init__` sets `self.n_cpu_tf_sess = n_cpu_tf_sess`.

[the remark this answers is not in the document]
Comment (shwang) [reply stable_baselines/acer/acer_simple.py]: I've moved this validation to the top of `ACER.__init__` by changing

```
        if num_procs is not None:
            warnings.warn(...)
            self.n_cpu_tf_sess = num_procs
```

to

```
       if num_procs is not None:
            warnings.warn(...)
            n_cpu_tf_sess = num_procs
```

File: stable_baselines/acktr/acktr.py (hunk 1 above)
Comment (shwang) [reply stable_baselines/acktr/acktr.py]: Addressed here: https://github.com/hill-a/stable-baselines/pull/639#discussion_r366134645

File: stable_baselines/common/base_class.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply stable_baselines/common/base_class.py]: OK, so the intention is that calling `self.runner` in these classes will raise a `NotImplementedError`, but there should be no such usage? Seems reasonable.

File: stable_baselines/acer/acer_simple.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply stable_baselines/acer/acer_simple.py]: :+1:

File: stable_baselines/common/identity_env.py [hunk 8]
```diff
@@ -5,22 +5,26 @@
...
         :param ep_length: (int) the length of each episodes in timesteps
         """
-        self.action_space = Discrete(dim)
-        self.observation_space = self.action_space
+        if space is None:
+            space = Discrete(dim)
+        self.action_space = self.observation_space = space
         self.ep_length = ep_length
         self.current_step = 0
         self.dim = dim
```
[the remark this answers is not in the document]
Comment (shwang) [reply stable_baselines/common/identity_env.py]: I don't understand, could you rephrase? (Where is what used?)

[the remark this answers is not in the document]
Comment (AdamGleave) [reply stable_baselines/common/identity_env.py]: I think I was just asking when `space is None` check hits, but I can see now that this is used in some other tests (I didn't have access to an IDE when I was writing this review).

File: stable_baselines/common/identity_env.py [hunk 9]
```diff
@@ -5,22 +5,26 @@
...
         """
         Identity environment for testing purposes
 
+        :param space: (gym.spaces.Space) The action and observation space.
         :param dim: (int) the size of the dimensions you want to learn
         :param ep_length: (int) the length of each episodes in timesteps
         """
-        self.action_space = Discrete(dim)
-        self.observation_space = self.action_space
+        if space is None:
```
[the remark this answers is not in the document]
Comment (shwang) [reply stable_baselines/common/identity_env.py]: Added explict check for exactly one None

File: stable_baselines/common/base_class.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shwang) [reply stable_baselines/common/base_class.py]: I don't see Runners being used in any way except to call `.run()` or accessing array shapes. (So no clean up happens AFAICT)

**Comment by AdamGleave** [comment]
> I'm removing the `IdentityBox.dim` attribute for now -- nothing is using it in our codebase and I'm guessing no else uses it either, if there's anyone using IdentityBox at all. Does this qualify as a breaking change? (@araffin)
> 
> Main reason I'm removing is because don't want to infer `dim` (always an integer, and now is an optional parameter to `IdentityBox.__init__`) from `space.shape` (sometimes an integer, sometimes tuple).

My gut instinct is we should document this in the changelog but overall not concerned by the change. I doubt `IdentityEnvBox` is used elsewhere and it explicitly says "identity environment for testing purposes", [GitHub](https://github.com/search?q=IdentityEnvBox&type=Code) search does not throw up any uses outside our codebase and forks. @araffin any thoughts?

**Comment by araffin** [comment]
yes, it's for testing only, so if the tests pass, no problem.

File: tests/test_multiple_learn.py (hunk 2 above)
Comment (AdamGleave) [reply tests/test_multiple_learn.py]: this doesn't seem to have been applied -- is that intentional?

File: stable_baselines/common/identity_env.py [hunk 10]
```diff
@@ -1,26 +1,41 @@
...
+        :param ep_length: the length of each episode in timesteps
         """
-        self.action_space = Discrete(dim)
-        self.observation_space = self.action_space
+        if space is None:
+            if dim is None:
+                dim = 1
+            space = Discrete(dim)
+        else:
+            assert dim is None
```
Comment (AdamGleave) [inline stable_baselines/common/identity_env.py]:
```suggestion
            assert dim is None, "arguments for both 'dim' and 'space' provided: at most one allowed"
```

**Comment by shwang** [comment]
@AdamGleave I only saw one comment where you referred to two, did you forget to confirm a review comment?

File: stable_baselines/common/identity_env.py [hunk 11]
```diff
@@ -1,26 +1,41 @@
...
+from gym import Env, Space
 from gym.spaces import Discrete, MultiDiscrete, MultiBinary, Box
 
 
 class IdentityEnv(Env):
-    def __init__(self, dim=1, ep_length=100):
+    def __init__(self,
+                 dim: Optional[int] = None,
+                 space: Optional[Space] = None,
+                 ep_length=100):
```
Comment (AdamGleave) [inline stable_baselines/common/identity_env.py]:
```suggestion
                 ep_length: int = 100):
```
